### PR TITLE
introduce param_groups to optimizer config

### PR DESCRIPTION
**Pull request by tianyu-l**
It is common to have different learning rates and weight decay rates for different parameters in the model.

Requested in https://github.com/pytorch/torchtitan/issues/2599, and as an alternative approach to https://github.com/pytorch/torchtitan/pull/2600.

In this PR we put `param_groups` as a `list[ParaGroupConfig]` into the optimizer config, to override lr and weight decay with multipliers based on regex matching.

### CODE REVIEW

**Comment by tianyu-l** [comment]
cc @francesco-bertolotti

**Comment by francesco-bertolotti** [comment]
This approach works well when the intended use is limited to overriding learning rate and weight decay for specific parameters. It would be great if other hyperparameters were also supported, such as `betas`. More generally, it would be even more powerful if the design allowed assigning entirely different optimizer types to different parameter groups, for example, using Muon for some parameters and AdamW for others.

For what it's worth, I've been working on exactly this in the latest revision of #2600, which introduces a `BaseOptimizer` class hierarchy with typed configs per optimizer, and supports mixed-optimizer parameter groups. 

Happy to discuss whether it makes sense to consolidate the two approaches!

**Comment by tianyu-l** [comment]
@francesco-bertolotti 

> More generally, it would be even more powerful if the design allowed assigning entirely different optimizer types to different parameter groups, for example, using Muon for some parameters and AdamW for others.

That makes sense (actually we have plans to support that).

Since the scope is larger than the original request in https://github.com/pytorch/torchtitan/issues/2599, maybe we can land this first (I'll add `betas` to the `ParamGroupConfig`) and address mixed optimizers in later PR.

**Review by wwwjn — APPROVED** [review]
